R/derive_date_vars.R

```diff
@@ -101,214 +117,730 @@
+#'  - For `date_imputation = "<mm>-<dd>"` `"xxxx"`, `"<mm>"`, `"<dd>"` are returned.
+#'
+#' @author Stefan Bundfuss
+#'
+#' @seealso [impute_dtc_dtm()], [impute_dtc_dt()]
+get_imputation_target_date <- function(date_imputation,
+                                       month) {
+  target <- vector("list", 3)
+  names(target) <- c("year", "month", "day")
+  if (date_imputation == "first") {
+    target[["year"]] <- "0000"
+    target[["month"]] <- "01"
+    target[["day"]] <- "01"
+  } else if (date_imputation == "mid") {
+    target[["year"]] <- "xxxx"
+    target[["month"]] <- "06"
+    target[["day"]] <- if_else(is.na(month), "30", "15")
+  } else if (date_imputation == "last") {
+    target[["year"]] <- "9999"
+    target[["month"]] <- "12"
+    target[["day"]] <- "28"
+  } else {
+    target[["year"]] <- "xxxx"
+    target[["month"]] <- str_sub(date_imputation, 1, 2)
+    target[["day"]] <- str_sub(date_imputation, 4, 5)
+  }
+  target
+}
 
-    # 3 blocks of  if/else statements that deal with date imputation and
-    # preserving partial dates.
-    # Ex: 2019---07 with MID and preserve = TRUE gives 2019-06-07
-    if (date_imputation == "mid" & preserve) {
-      imputed_date <- case_when(
-        n_chr == 9 ~ paste0(substr(dtc, 1, 4), "-", "06", "-", substr(dtc, 8, 9)),
-        n_chr == 4 ~ paste0(dtc, "-", "06", "-", "30"),
-        TRUE ~ imputed_date
+#' Get Time Imputation Targets
+#'
+#' @param time_imputation The value to impute the time when a timepart is
+#'   missing.
+#'
+#'   A character value is expected, either as a
+#'   - format with hour, min and sec specified as `"hh:mm:ss"`: e.g. `"00:00:00"`
+#'   for the start of the day,
+#'   - or as a keyword: `"first"`,`"last"` to impute to the start/end of a day.
+#'
+#' @returns A list of character vectors. The elements of the list are named
+#'   "hour", "minute", "second".
+#'
+#' @details
+#'
+#'  - For `time_imputation = "first"` `"00"`, `"00"`, `"00"` are returned.
+#'  - For `time_imputation = "last"` `"23"`, `"59"`, `"59"` are returned.
+#'  - For `time_imputation = "<hh>:<mm>:<ss>"` `"<hh>"`, `"<mm>"`, `"<ss>"` are returned.
+#'
+#' @author Stefan Bundfuss
+#'
+#' @seealso  [impute_dtc_dtm()]
+get_imputation_target_time <- function(time_imputation) {
+  target <- vector("list", 3)
+  names(target) <- c("hour", "minute", "second")
+  if (time_imputation == "first") {
+    target[["hour"]] <- "00"
+    target[["minute"]] <- "00"
+    target[["second"]] <- "00"
+  } else if (time_imputation == "last") {
+    target[["hour"]] <- "23"
+    target[["minute"]] <- "59"
+    target[["second"]] <- "59"
+  } else {
+    target[["hour"]] <- str_sub(time_imputation, 1, 2)
+    target[["minute"]] <- str_sub(time_imputation, 4, 5)
+    target[["second"]] <- str_sub(time_imputation, 7, -1)
+  }
+  target
+}
+
+#' Restrict Imputed DTC date to Minimum/Maximum Dates
+#'
+#' @param imputed_dtc The imputed DTC date
+#'
+#' @inheritParams impute_dtc_dtm
+#'
```

Comment:
Does @inheritParams retrieve keywords and families as well?  If not, missing the keywords and families.

Reply:
No, keyword and family added.